### PR TITLE
HIVE-22415: Upgrade to Java 11

### PR DESCRIPTION
**Pull request by belugabehr**

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/resources/datanucleus-log4j.properties [hunk 1]
```diff
@@ -15,3 +15,5 @@ log4j.category.DataNucleus.ValueGeneration=DEBUG, A1
 
 log4j.category.DataNucleus.Enhancer=INFO, A1
 log4j.category.DataNucleus.SchemaTool=INFO, A1
+
+log4j.category.DataNucleus.Persistence=INFO, A1
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/resources/datanucleus-log4j.properties:19]: Remove this change.

**Comment by kgyrtkirk** [comment]
oh...I see kgyrtkirk/hive-dev-box#5 is needed to test this

**Comment by kgyrtkirk** [comment]
how about we phasing in java 8 + java 11 testing at the same time?
@belugabehr  would that make sense?

**Comment by belugabehr** [comment]
@kgyrtkirk 

Yup!  OpenJDK and Oracle are the only ones Cloudera supports.  For Java 11, it's OpenJDK 11 only.

https://docs.cloudera.com/cdp/latest/release-guide/topics/cdpdc-java-requirements.html

**Comment by belugabehr** [comment]
@kgyrtkirk I'm not sure what you mean by `java 8 + java 11 testing`.  If we go with JDK 11 for the build, it should be backwards compatible with Java 8 (which is defined in the Maven POM file for Hive 3.x and master currently).

File: standalone-metastore/metastore-server/pom.xml [hunk 2]
```diff
@@ -726,8 +726,7 @@
         <configuration>
           <api>JDO</api>
           <verbose>false</verbose>
-          <log4jConfiguration>${basedir}/src/main/resources/datanucleus-log4j.properties
-          </log4jConfiguration>
+          <log4jConfiguration>${basedir}/src/main/resources/datanucleus-log4j.properties</log4jConfiguration>
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/pom.xml:729]: Remove this change.

File: .travis.yml [hunk 3]
```diff
@@ -24,17 +23,14 @@ dist: bionic
...
+  - openjdk11
 
 env:
   MAVEN_SKIP_RC=true
   MAVEN_OPTS="-Xmx2g"
 
 install: skip
 
-script: mvn clean install -DskipTests -q -Pitests
+before_script: mvn --version
```
Comment (kgyrtkirk) [inline .travis.yml:34]: I think we should just remove `travis` entirely...no need to do this 2 times

Comment (belugabehr) [reply .travis.yml:34]: What is the difference between the two exactly?  I like that I can easily create per-branch build envs with Travis.

**Comment by kgyrtkirk** [comment]
you could switch everything to jdk 11 by adding
```
sw java 11
```
around here:
https://github.com/apache/hive/blob/e34acf5c677a23af0053ac98532a9caa9e190b6c/Jenkinsfile#L84

but I would prefer to keep jdk8 compatibility - wouldn't that be possible?

in case it's harder to do everything in 1 round ( I suspect we will need a new Tez version as well) we can get in a `compile on jdk11 as well` patch first and continue from there

**Comment by belugabehr** [comment]
Once the project is compiled in JDK 11, it will no longer run in JRE 8.  Something like Tez is a dependency: JRE 11 can run something compiled with JDK 8.  So, if a dependency is compiled with an older version of the compiler, that's not a problem.

**Comment by kgyrtkirk** [comment]
iirc you won't be able to run the tests unles tez is jdk11 compliant

File: .travis.yml (hunk 3 above)
Comment (kgyrtkirk) [reply .travis.yml:34]: travis doesn't run tests - it just compiles

File: llap-server/src/java/org/apache/hadoop/hive/llap/cache/SimpleAllocator.java [hunk 4]
```diff
@@ -17,31 +17,17 @@
...
 import java.nio.ByteBuffer;
 
 import org.apache.hadoop.conf.Configuration;
 import org.apache.hadoop.hive.common.io.Allocator;
 import org.apache.hadoop.hive.common.io.encoded.MemoryBuffer;
 import org.apache.hadoop.hive.conf.HiveConf;
 import org.apache.hadoop.hive.llap.io.api.impl.LlapIoImpl;
-
-import sun.misc.Cleaner;
+import org.apache.hive.common.util.CleanerUtil;
```
Comment (kgyrtkirk) [inline llap-server/src/java/org/apache/hadoop/hive/llap/cache/SimpleAllocator.java:27]: why don't we use the `CleanerUtil` from hadoop? I don't think we should copy it

https://issues.apache.org/jira/browse/HADOOP-12760

Comment (belugabehr) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/SimpleAllocator.java:27]: @kgyrtkirk I would prefer this too, but this class was made available in Hadoop 3.2.0 and Hive is on Hadoop 3.1.0

Comment (kgyrtkirk) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/SimpleAllocator.java:27]: I think we should upgrade to hadoop >=3.2.0 (in a separate ticket) prior to moving to jdk11...

Comment (belugabehr) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/SimpleAllocator.java:27]: Other than saving a copy & paste on this one class, I'm not sure the value.  I don't think it should be a blocker on this.

Comment (kgyrtkirk) [reply llap-server/src/java/org/apache/hadoop/hive/llap/cache/SimpleAllocator.java:27]: I think it's mandatory to upgrade to at least
https://issues.apache.org/jira/browse/HADOOP-15338 marks 3.3 for jdk11 support - I don't think we should try to run hive with hadoop 3.1 on jdk11...

File: .travis.yml (hunk 3 above)
Comment (belugabehr) [reply .travis.yml:34]: Ya, sorry, I mean from a feature standpoint : Travis v.s. Jenkins

**Comment by belugabehr** [comment]
@kgyrtkirk Let the build run and see what we get.  Again, it's different to say that something compiles under JDK11 and something runs.  Also, since Hive is a client application, the dependencies may not have experienced any changes.

I am not against an upgrade of Hadoop library.  We just don't need to link these two efforts unless we absolutely have to.

File: Jenkinsfile [hunk 5]
```diff
@@ -81,6 +81,7 @@ def buildHive(args) {
   configFileProvider([configFile(fileId: 'artifactory', variable: 'SETTINGS')]) {
     withEnv(["MULTIPLIER=$params.MULTIPLIER","M_OPTS=$params.OPTS"]) {
       sh '''#!/bin/bash -e
+sw java 11
```
Comment (belugabehr) [inline Jenkinsfile:84]: @kgyrtkirk 
```
Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.1:compile (default-compile) on project hive-storage-api: Fatal error compiling: invalid target release: 11 -> [Help 1]
```
This didn't seem to work.

http://ci.hive.apache.org/blue/organizations/jenkins/hive-precommit/detail/PR-1241/5/pipeline

**Comment by ashutoshc** [comment]
My suggestion would be to split the changes in separate tickets. Ticket for code changes which are necessary for jdk11 but can be committed before the actual move to jdk11 like hadoop class/version, dependency exclusion, dependency upgrades etc.  Commit those first and then have minimal changes after that just changes the compile and runtime version to 11. Advantage of doing it that way is it will make it easier downstream to cherry-pick up these changes if they are separate commits.

**Comment by belugabehr** [comment]
@ashutoshc  Happy to do that.  Let me get it all working in this PR then I can start cutting into phases.

**Comment by belugabehr** [comment]
@kgyrtkirk Cannot build `init-metastore` part of the pipeline.  Does this somehow use a different docker image?  How do I get it's `JAVA_HOME` to JDK 11?

**Comment by abstractdog** [comment]
FYI, HIVE-23930 is resolved, hive now depends on tez 0.10.0 which is JDK11 compliant

**Comment by belugabehr** [comment]
@abstractdog Re-opened PR at #1624

**Pull request closed without merging**